### PR TITLE
[Object_Detection] Add Hard Negative Mining.

### PR DESCRIPTION
**Pull request by tanzhenyu**
Note that this does NOT implement the caveat described in
[Training Region-based Object Detectors with Online Hard Example Mining]
(https://arxiv.org/pdf/1604.03540.pdf), section 4.1, where NMS is used
to perform deduplication. I believe this can be done before getting to
this layer, but it can also be done as part of this layer. For now,
I will prefer composibility. If NMS is ALWAYS used together with negative
mining, then we should consider adding an argument in this layer.

### CODE REVIEW

**Comment by bhack** [comment]
@tanzhenyu Just a memory from the past if you are interested https://github.com/tensorflow/tensorflow/issues/17693#issuecomment-378676163

**Comment by tanzhenyu** [comment]
> @tanzhenyu Just a memory from the past if you are interested [tensorflow/tensorflow#17693 (comment)](https://github.com/tensorflow/tensorflow/issues/17693#issuecomment-378676163)

Aha, thanks!
Meanwhile I'm not sure what we can do for the "online" piece. What we're doing here is batched version and mostly for offline.

**Comment by bhack** [comment]
> Meanwhile I'm not sure what we can do for the "online" piece. 
Yes this is still the interesting part of the topic

**Comment by tanzhenyu** [comment]
> > Meanwhile I'm not sure what we can do for the "online" piece.
> 
> Yes this is still the interesting part of the topic

The entire online training with TF / Keras is not my expertise (or I'm not aware that it supports it at all)

**Comment by bhack** [comment]
Yes was a more articulated topic that didn't find enough attention in the original ticket.

File: kerascv/layers/losses/hard_neg_miner.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import tensorflow as tf
+
+
+class HardNegativeMiner(tf.keras.layers.Layer):
+    """Hard Negative Mining to keep balanced positive and negative samples."""
```
Comment (fchollet) [inline kerascv/layers/losses/hard_neg_miner.py:5]: Please add a more detailed docstring, including information about the constructor arguments, call inputs, and call return value

File: kerascv/layers/losses/hard_neg_miner.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import tensorflow as tf
+
+
+class HardNegativeMiner(tf.keras.layers.Layer):
```
Comment (fchollet) [inline kerascv/layers/losses/hard_neg_miner.py:4]: "HardNegativeMining" would be more in line with the usual layer naming conventions (e.g. `Masking`)

File: kerascv/layers/losses/hard_neg_miner.py (hunk 1 above)
Comment (tanzhenyu) [reply kerascv/layers/losses/hard_neg_miner.py]: Done.

File: kerascv/layers/losses/hard_neg_miner.py (hunk 2 above)
Comment (tanzhenyu) [reply kerascv/layers/losses/hard_neg_miner.py]: Good point. Done.

**Comment by tanzhenyu** [comment]
> Thank you for the PR. Should this logic be shared with hard positive mining as well? Should we expose a `HardPositiveMining` layer?

What is `HardPositiveMining`?

**Comment by fchollet** [comment]
Hard positive mining is the positive counterpart to hard negative mining. It should be very much the same logic. See e.g. [section 3.4 here](http://openaccess.thecvf.com/content_ECCV_2018/papers/SouYoung_Jin_Unsupervised_Hard-Negative_Mining_ECCV_2018_paper.pdf)

My comment is just a question -- something to keep in mind. If you don't think it's actionable, no need to take action.

**Comment by tanzhenyu** [comment]
> Hard positive mining is the positive counterpart to hard negative mining. It should be very much the same logic. See e.g. [section 3.4 here](http://openaccess.thecvf.com/content_ECCV_2018/papers/SouYoung_Jin_Unsupervised_Hard-Negative_Mining_ECCV_2018_paper.pdf)
> 
> My comment is just a question -- something to keep in mind. If you don't think it's actionable, no need to take action.

Ah, thanks! Maybe we could consider this by having 'minimum_positive_examples' and sort on positive values in the future
For now I think this is probably not widely used enough to be part of Keras-CV.

**Comment by bhack** [comment]
See also hard positives in the context of:

https://arxiv.org/abs/1909.09777
(Official impl) https://github.com/kemaloksuz/BoundingBoxGenerator

**Comment by tanzhenyu** [comment]
> See also hard positives in the context of:
> 
> https://arxiv.org/abs/1909.09777
> (Official impl) https://github.com/kemaloksuz/BoundingBoxGenerator

This particular paper (I haven't read through) seems to be done at the anchor generator / roi generator level, instead of the loss mining level

**Comment by bhack** [comment]
Yes

**Comment by tanzhenyu** [comment]
> Yes

But very interesting :-) Thanks for pointing it out!

File: kerascv/layers/losses/hard_neg_miner.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        negative_positive_ratio: An int to represent the maximum ratio between
+            negative samples and positive samples. For example, if the inputs
+            has negative samples : positive samples = 100 : 20, it will sort all
+            negative samples by `values` and only keep the top 60. Defaults to `3`.
+        minimum_negative_examples: An int to represent the minimum negative examples
+            to have. For example if negative samples = 100, positive samples = 20,
+            `negative_positive_ratio=1` and `minimum_negative_examples=30`, it will
+            keep the top 30 results.
+
+    Inputs:
```
Comment (fchollet) [inline kerascv/layers/losses/hard_neg_miner.py:20]: This is usually called "Call arguments:"
